sensorium/training/trainers.py

```diff
@@ -188,21 +311,116 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
 
             batch_args = list(data)
             batch_kwargs = data._asdict() if not isinstance(data, dict) else data
-            loss = full_objective(
+            loss, loss_parts = full_objective(
                 model,
                 dataloaders["train"],
                 data_key,
                 *batch_args,
                 **batch_kwargs,
-                detach_core=detach_core
+                detach_core=detach_core,
             )
             loss.backward()
+            epoch_loss += loss.detach()
+            epoch_loss_main += loss_parts[0].detach()
+            epoch_loss_reg += loss_parts[1].detach()
+            cluster_centers_list = []
             if (batch_no + 1) % optim_step_count == 0:
+                # TODO maybe remove the hidden dimensions
+                if include_kldivergence and epoch >= dec_starting_epoch:
+                    kldiv_loss = torch.zeros(1).to(device)
+                    features_subset = []
+                    feature_list = []
+                    for k, readout in model.readout.items():
+                        features = readout.features.detach().squeeze()
```

Comment:
why do you do `detach` here?
I would suggest to do `features = readout.features.squeeze()` because otherwise this might mess up gradient [computations](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html#)

Reply:
I changed that

Reply:
if you commit and push to the same branch (`[KLdiv_branch`) I should see the changes here (the code in PR is automatically updated once you push to the source branch)

Reply:
but better do it after addressing all the comments:)